src/runtime-rs/crates/hypervisor/src/dragonball/inner.rs

```diff
@@ -0,0 +1,300 @@
+
+        let mut abs_path = String::from("/");
+        abs_path.push_str(dst);
+        Ok(abs_path)
+    }
+
+    fn set_boot_source(&mut self, kernel_path: &str, kernel_params: &str) -> Result<()> {
+        info!(
+            sl!(),
+            "kernel path {} kernel params {}", kernel_path, kernel_params
+        );
+
+        let mut boot_cfg = BootSourceConfig {
+            kernel_path: self
+                .get_resource(kernel_path, DRAGONBALL_KERNEL)
+                .context("jail resource")?,
+            ..Default::default()
+        };
+
+        if !kernel_params.is_empty() {
+            boot_cfg.boot_args = Some(kernel_params.to_string());
+        }
+
+        self.vmm_instance
+            .put_boot_source(boot_cfg)
+            .context("put boot source")
+    }
+
+    fn set_vm_rootfs(&mut self, path: &str, driver: &str) -> Result<()> {
+        info!(sl!(), "==== set vm rootfs {} {}", path, driver);
+        let jail_drive = self
+            .get_resource(path, DRAGONBALL_ROOT_FS)
+            .context("get resource")?;
+
+        if driver == VM_ROOTFS_DRIVER_BLK {
+            let blk_cfg = BlockDeviceConfigInfo {
+                path_on_host: PathBuf::from(jail_drive),
+                drive_id: DRAGONBALL_ROOT_FS.to_string(),
+                is_root_device: false,
+                // Add it as a regular block device
+                // This allows us to use a partitioned root block device
+                // is_read_only
+                is_read_only: true,
+                is_direct: false,
+                ..Default::default()
+            };
+
+            self.vmm_instance
+                .insert_block_device(blk_cfg)
+                .context("inert block device")
+        } else {
+            Err(anyhow!("Unknown vm_rootfs driver: {}", driver))
+        }
+    }
+
+    fn start_vmm_instance(&mut self) -> Result<()> {
+        info!(sl!(), "Starting VM");
+        self.vmm_instance
+            .instance_start()
+            .context("Failed to start vmm")?;
+        self.state = VmmState::VmRunning;
+        Ok(())
+    }
+
+    // wait_vmm will wait for timeout seconds for the VMM to be up and running.
+    // This does not mean that the VM is up and running. It only indicates that the VMM is up and
+    // running and able to handle commands to setup and launch a VM
+    fn wait_vmm_ready(&mut self, timeout: i32) -> Result<()> {
+        if timeout < 0 {
+            return Err(anyhow!("Invalid param timeout {}", timeout));
+        }
+
+        let time_start = std::time::Instant::now();
+        loop {
+            match self.vmm_instance.is_running() {
+                Ok(_) => return Ok(()),
+                Err(err) => {
+                    let time_now = std::time::Instant::now();
+                    if time_now.duration_since(time_start).as_millis() > timeout as u128 {
+                        return Err(anyhow!(
+                            "Describe instance timeout {} err : {:?}",
```

Comment:
Nit: though internally `Describe` API may be called, here `Describe` is not very clear.
How about `waiting vmm ready timeout`?

Reply:
Thanks for your review, already updated.